[AArch64][SVE] Add codegen support for partial reduction lowering to wide add instructions

JamesChesterman opened the pull request:
For partial reductions in the situation of the number of lanes being halved and the bits per lane being doubled, a pair of wide add instructions can be used.

JamesChesterman committed:
[AArch64][SVE] Add codegen support for partial reduction lowering to wide add instructions

huntergr-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -2042,7 +2042,8 @@ bool AArch64TargetLowering::shouldExpandPartialReductionIntrinsic(
   EVT VT = EVT::getEVT(I->getType());
   return VT != MVT::nxv4i64 && VT != MVT::nxv4i32 && VT != MVT::nxv2i64 &&
-         VT != MVT::v4i64 && VT != MVT::v4i32 && VT != MVT::v2i32;
+         VT != MVT::nxv8i16 && VT != MVT::v4i64 && VT != MVT::v4i32 &&
```
I think we may want to start looking at the inputs at this point, as the types that we have instructions for differ based on the chain of operations on the second operand. May be something for a follow-up patch.

JamesChesterman replied:
Done

MacDue commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -2042,7 +2042,8 @@ bool AArch64TargetLowering::shouldExpandPartialReductionIntrinsic(
 
   EVT VT = EVT::getEVT(I->getType());
   return VT != MVT::nxv4i64 && VT != MVT::nxv4i32 && VT != MVT::nxv2i64 &&
-         VT != MVT::v4i64 && VT != MVT::v4i32 && VT != MVT::v2i32;
+         VT != MVT::nxv8i16 && VT != MVT::v4i64 && VT != MVT::v4i32 &&
+         VT != MVT::v2i32 && VT != MVT::v8i16;
 }
```
I know this was already like this, but I feel like an `is_contained()` check would be more readable than these big expressions :thinking:  
```c++
  EVT VT = EVT::getEVT(I->getType());
  return !is_contained<EVT>({MVT::nxv4i64, MVT::nxv4i32, MVT::nxv2i64,
                             MVT::nxv8i16, MVT::v4i64, MVT::v4i32, MVT::v2i32,
                             MVT::v8i16},
                            VT);
```

JamesChesterman replied:
Refactored this code so this isn't applicable anymore. I'll bear this function in mind for future development though.

huntergr-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -21783,6 +21784,62 @@ SDValue tryLowerPartialReductionToDot(SDNode *N,
+  // To do this transformation, output element size needs to be double input
+  // element size, and output number of elements needs to be half the input
+  // number of elements
+  if (!(InputType.getVectorElementType().getSizeInBits() * 2 ==
```
Can remove the `!(a == b)` in favour of `a != b`. `||` is lower precedence so we don't need extra parens here.

JamesChesterman replied:
Done

MacDue commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -21783,6 +21784,62 @@ SDValue tryLowerPartialReductionToDot(SDNode *N,
+
+  SDLoc DL(N);
+
+  auto Accumulator = N->getOperand(1);
+  auto ExtInput = N->getOperand(2);
+
+  EVT AccumulatorType = Accumulator.getValueType();
+  EVT AccumulatorElementType = AccumulatorType.getVectorElementType();
```
nit (optional), I think these names could use some comment shorthands:

- `Accumulator` -> `Acc`
- `AccumulatorType` -> `AccVT`
- `AccumulatorElementType` -> `AccElemVT`
- `InputType` -> `InputVT`

JamesChesterman replied:
Done

huntergr-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -21783,6 +21784,62 @@ SDValue tryLowerPartialReductionToDot(SDNode *N,
+    return SDValue();
+
+  bool InputIsSigned = ExtInputOpcode == ISD::SIGN_EXTEND;
+  auto BottomIntrinsic = InputIsSigned ? Intrinsic::aarch64_sve_saddwb
```
I think adding dedicated AArch64ISD nodes for these operations would be good, though that can also be in another patch.

JamesChesterman replied:
Will be done in a future PR

huntergr-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -21783,6 +21784,62 @@ SDValue tryLowerPartialReductionToDot(SDNode *N,
+         "Expected a partial reduction node");
+
+  bool Scalable = N->getValueType(0).isScalableVector();
+  if (Scalable && !Subtarget->isSVEorStreamingSVEAvailable())
```
Based on the intrinsics you're lowering to below, I think you need to bail if SVE isn't available regardless. There are similar instructions for NEON but they aren't matched here.

JamesChesterman replied:
Done

JamesChesterman committed:
Minor changes to previous patch

Rename variables, eliminate a redundant condition in an if
statement and refactor code checking types

MacDue commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -2041,8 +2041,13 @@ bool AArch64TargetLowering::shouldExpandPartialReductionIntrinsic(
-         VT != MVT::v4i64 && VT != MVT::v4i32 && VT != MVT::v2i32;
+  auto Op1 = I->getOperand(1);
+  EVT Op1VT = EVT::getEVT(Op1->getType());
+  if (Op1VT.getVectorElementType() == VT.getVectorElementType() &&
+      (VT.getVectorElementCount() * 4 == Op1VT.getVectorElementCount() ||
+       VT.getVectorElementCount() * 2 == Op1VT.getVectorElementCount()))
```
The previous check limited this to specific types, this only looks at the element counts, but presumably not all of these can be mapped to instructions. What happens int those cases?

MacDue replied:
It looks like it should fallback to `DAG.getPartialReduceAdd()` (in `performIntrinsicCombine()`), but `tryLowerPartialReductionToWideAdd()` does not seem to check if there is a wide add instruction available.

JamesChesterman replied:
I did this with the view of implementing type legalisation I think. For example, we do want a partial reduction from `nxv4i16` to `nxv2i32` (after the input has been extended) to be possible right? And I thought that all that would be needed to be done in checking the wide add instruction is available is checking whether SVE was enabled, or is there something I'm missing?

JamesChesterman replied:
I think you may be right, it doesn't seem like other intrinsics like dot product have legalisation. So do you think that all the tests required are already in `sve-partial-reduce-wide-add.ll` ? `(`nxv4i32`->nxv2i64`, `nxv8i16`->`nxv4i32`, `nxv16i8`->`nxv8i16`)

JamesChesterman replied:
This is now resolved with an additional check in the `tryLowerPartialReductionToWideAdd` function.

MacDue replied:
Cool, thanks!

JamesChesterman committed:
Check for non-legal types for the wide add instructions

Check for non-legal types for the wide add instructions so that
instructions that require promotion or splitting just go to the
default case.

MacDue commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -21783,6 +21788,64 @@ SDValue tryLowerPartialReductionToDot(SDNode *N,
+  // element size, and output number of elements needs to be half the input
+  // number of elements
+  if (InputVT.getVectorElementType().getSizeInBits() * 2 !=
+          AccElemVT.getSizeInBits() ||
+      AccVT.getVectorElementCount() * 2 != InputVT.getVectorElementCount() ||
```
nit: I think this check is redundant with the type checks above.

JamesChesterman replied:
I just thought I'd leave it here ready for the future legalisation work. But I can get rid of it if you like and just put it back in with the future patch?

MacDue replied:
If the legalization is done before lowering the partial reduction you should have legal types here. I don't think the legalization should be done on the `AArch64::SADDWB` ISD node for example, as I think those would be expected to have 1-to-1 mapping with the instructions.

JamesChesterman replied:
Done

MacDue reviewed (approved):
LGTM (just one nit). I'd wait for @huntergr-arm to approve before landing though :slightly_smiling_face:

JamesChesterman committed:
Remove unnecessary type checking

huntergr-arm reviewed (approved):
LG
